### PR TITLE
Major revision of the Scifi digitization

### PR DESCRIPTION
**Pull request by siilieva**
New lab measurements have been performed that led to an updated version of the Scifi digitization.
It is being described in an internal note.
Completely removed the old model since it's results are much more different from the data. 
For reference the presentations describing the comparisons between MC Scifi realizations and data are given below
https://indico.cern.ch/event/1394068/contributions/5893765/attachments/2830724/4945770/Software_news_3April2024.pdf
https://indico.cern.ch/event/1410259/contributions/5945569/attachments/2852834/4988435/Software_news_8May2024_new_Scifi_digi.pdf

### CODE REVIEW

File: shipLHC/sndScifiHit.cxx [hunk 1]
```diff
@@ -103,10 +101,22 @@ Float_t sndScifiHit::GetEnergy()
   return signals[0];
 }
 
-Float_t sndScifiHit::ly_loss_mean(Float_t distance, Float_t* params){
-//	It returns the light yield depending on the distance to SiPM
-	Double_t att = params[0] * TMath::Exp(params[1] * distance / 100.) + params[2] * TMath::Exp(params[3] * distance / 100.);
-	return att/(params[0]+params[2]);
+Float_t sndScifiHit::ly_loss(Float_t distance, Float_t* params){
```
Comment (olantwin) [inline shipLHC/sndScifiHit.cxx:104]: 1. `light_yield_loss` or `light_attenuation`?
2. Could we please use a fixed-size array for params? It seems it is expected to be length 2?

File: shipLHC/sndScifiHit.cxx [hunk 2]
```diff
@@ -103,10 +101,22 @@ Float_t sndScifiHit::GetEnergy()
...
+//	It returns the light yield attenuation depending on the distance to SiPM
+	return TMath::Exp(-(distance-params[0])/params[1]);
+}
+
+Float_t sndScifiHit::sipm_saturation(Float_t ly, Float_t nphe_max){
```
Comment (olantwin) [inline shipLHC/sndScifiHit.cxx:109]: Based on the comment, I would expect the return value to be an integer?

File: shipLHC/sndScifiHit.cxx [hunk 3]
```diff
@@ -103,10 +101,22 @@ Float_t sndScifiHit::GetEnergy()
...
+//	It returns the number of fired pixels per channel
+        Float_t factor = 1 - TMath::Exp(-ly/nphe_max);
+	return nphe_max*factor;
+}
+
+Float_t sndScifiHit::npix_to_qdc(Float_t npix, Float_t* params){
```
Comment (olantwin) [inline shipLHC/sndScifiHit.cxx:115]: As above, let's use fixed-length arrays.

File: shipLHC/sndScifiHit.cxx [hunk 4]
```diff
@@ -73,16 +61,26 @@ sndScifiHit::sndScifiHit (int SiPMChan, std::vector<ScifiPoint*> V, std::vector<
...
 	
      }
-     Float_t ly = signalTotal/0.180*1000.*20.; // 20 p.e. per 180 keV
-     signals[0] =MeanAndRMS(ly,nphe_max);
-     if (ly >nphe_min){   // nominal threshold at 3.5 p.e.
+     // smear the total light yield using Poisson distribution
+     ly_total= gRandom->Poisson(ly_total);
+
+     // account for limited SiPM dyn. range
+     Float_t Npix = sipm_saturation(ly_total,nphe_max);
```
Comment (olantwin) [inline shipLHC/sndScifiHit.cxx:79]: Why is this not an integer?

File: shipLHC/sndScifiHit.h [hunk 5]
```diff
@@ -42,14 +42,16 @@ class sndScifiHit : public SndlhcHit
...
-    Float_t MeanAndRMS(Double_t ly, Float_t T);
+    Float_t ly_loss(Float_t distance, Float_t* params);
+    Float_t sipm_saturation(Float_t ly, Float_t nphe_max);
+    Float_t npix_to_qdc(Float_t npix, Float_t* params);
     Float_t flag;   ///< flag
 
 // get parameters from the Scifi detector for simulating the digitized information
-	Float_t ly_loss_params[4] = {20.78, -0.26, 7.89, -3.06};
+	Float_t ly_loss_params[4] = {20., 300.}; //x_0, lambda
+	Float_t npix_to_qdc_params[4] = {0.172, -1.31, 0.006, 0.33}; // A, B, sigma_A, sigma_B
```
Comment (olantwin) [inline shipLHC/sndScifiHit.h:52]: These should probably be static as they appear to be global constants?
In which case, I don't see the point of passing them as arguments to the different functions.

**Review by olantwin — CHANGES_REQUESTED** [review]
1. I think some of the names are unnecessarily cryptic.
2. If the parameters are fixed anyway, we should probably make them static constants. We might also want to make sure that they are not serialised. Having a mutable copy per hit seems unnecessary. This would also simplify the method signatures.

File: shipLHC/sndScifiHit.cxx [hunk 6]
```diff
@@ -73,16 +61,26 @@ sndScifiHit::sndScifiHit (int SiPMChan, std::vector<ScifiPoint*> V, std::vector<
 	  distance = (impact - a).Mag();
 	}
 
-	signalTotal+= signal*ly_loss_mean(distance,ly_loss_params);
+	// convert energy deposit to light yield (here Np.e. == avg. N fired pixels)
+        Float_t ly = signal*1E+6*0.16; //0.16 p.e per 1 keV
```
[the remark this answers is not in the document]
Comment (siilieva) [reply shipLHC/sndScifiHit.cxx:65]: `ly` is quite common abbreviation for light yield
If the future reader/user is in doubt the comment above explicitly reads "light yield"

File: shipLHC/sndScifiHit.cxx (hunk 1 above)
Comment (siilieva) [reply shipLHC/sndScifiHit.cxx:104]: same for the ly_loss as above
2. good catch! I did a little research and it is suggested that the best practice for such internal class constants is an anonymous namespace, so I did that.

File: shipLHC/sndScifiHit.cxx (hunk 2 above)
Comment (siilieva) [reply shipLHC/sndScifiHit.cxx:109]: it could be non integer, same as the threshold in photo electrons is 3.5

File: shipLHC/sndScifiHit.cxx (hunk 3 above)
Comment (siilieva) [reply shipLHC/sndScifiHit.cxx:115]: anonymous namespace introduced.

File: shipLHC/sndScifiHit.cxx (hunk 4 above)
Comment (siilieva) [reply shipLHC/sndScifiHit.cxx:79]: same as above
These are all internal calculations to get the QDC - better have non integer Npix etc. than lose precision of the final QDC.

File: shipLHC/sndScifiHit.h (hunk 5 above)
Comment (siilieva) [reply shipLHC/sndScifiHit.h:52]: these are the fixed length params - placed in a namespace after the revision

File: shipLHC/sndScifiHit.cxx [hunk 7]
```diff
@@ -8,16 +8,10 @@
...
-     Double_t RMS = -0.00055223   + 0.0667554*ly
-                       -0.000478245*ly*ly + 1.48809e-06*pow(ly,3)
-                       -2.26064e-09*pow(ly,4) + 1.35786e-12*pow(ly,5);
-     Float_t r = gRandom->Gaus(mean, RMS);
-     Float_t signal = std::min(nphe_max,r);
-     return signal;
+namespace {
+     //parameters for simulating the digitized information
+     Float_t ly_loss_params[4] = {20., 300.}; //x_0, lambda
+     Float_t npix_to_qdc_params[4] = {0.172, -1.31, 0.006, 0.33}; // A, B, sigma_A, sigma_B
```
Comment (olantwin) [inline shipLHC/sndScifiHit.cxx:14]: We might want to make these constant.

**Review by olantwin — DISMISSED** [review]
Looks good, the anonymous namespace makes things a lot tidier.

File: shipLHC/sndScifiHit.cxx (hunk 7 above)
Comment (siilieva) [reply shipLHC/sndScifiHit.cxx:14]: absolutely! I forgot to do it in the previous edit - doing now

**Review by olantwin — DISMISSED** [review]

**Comment by siilieva** [comment]
I believe another review is needed since there is a new commit  :innocent:

**Review by olantwin — APPROVED** [review]
I like that change to the copy constructor and assignment operator a lot!

**Pull request merged**